nixos/modules/services/x11/extra-layouts.nix

```diff
@@ -0,0 +1,232 @@
+          <vendor>${layout.description}</vendor>
+        </configItem>
+      </model>
+      .
+      w
+      EOF
+
+      # add layout description
+      ${pkgs.ed}/bin/ed -v rules/base.xml <<EOF
+      /<\/layoutList>
+      -
+      a
+      <layout>
+        <configItem>
+          <name>${name}</name>
+          <_shortDescription>${name}</_shortDescription>
+          <_description>${layout.description}</_description>
+          <languageList>
+            ${lib.concatMapStrings (lang: "<iso639Id>${lang}</iso639Id>\n") layout.languages}
+          </languageList>
+        </configItem>
+        <variantList/>
+      </layout>
+      .
+      w
+      EOF
+  '';
+
+in
+
+{
+
+  ###### interface
+
+  options.services.xserver = {
+    extraLayouts = mkOption {
+      type = types.attrsOf (types.submodule layoutOpts);
+      default = {};
+      example = literalExample
+      ''
+        {
+          mine = {
+            description = "My custom xkb layout.";
+            languages = [ "eng" ];
+            symbolsFile = /path/to/my/layout;
+          };
+        }
+      '';
+      description = ''
+        Extra custom layouts that will be included in the xkb configuration.
+        Information on how to create a new layout can be found here:
+        https://www.x.org/releases/current/doc/xorg-docs/input/XKB-Enhancing.html#Defining_New_Layouts
+        For more examples see:
+        https://wiki.archlinux.org/index.php/X_KeyBoard_extension#Basic_examples
+      '';
+    };
+
+  };
+
+  ###### implementation
+
+  config = mkIf (layouts != { }) {
+
+    # We don't override xkeyboard_config directly to
+    # reduce the amount of packages to be recompiled.
+    # Only the following packages are necessary to set
+    # a custom layout anyway:
+    nixpkgs.overlays = lib.singleton (self: super: {
+
+      xkb_patched = self.xorg.xkeyboardconfig.overrideAttrs (old: {
+        buildInputs = old.buildInputs ++ [ self.automake ];
+        postPatch = with lib; concatStrings (mapAttrsToList (patchIn self) layouts);
+      });
+
+      xorg = super.xorg // {
+        xorgserver = super.xorg.xorgserver.overrideAttrs (old: {
+          configureFlags = old.configureFlags ++ [
+            "--with-xkb-bin-directory=${self.xorg.xkbcomp}/bin"
+            "--with-xkb-path=${self.xkb_patched}/share/X11/xkb"
+          ];
+        });
+
+        setxkbmap = super.xorg.setxkbmap.overrideAttrs (old: {
+          postInstall =
+            ''
+              mkdir -p $out/share
+              ln -sfn ${self.xkb_patched}/etc/X11 $out/share/X11
+            '';
+        });
+
+        xkbcomp = super.xorg.xkbcomp.overrideAttrs (old: {
+          configureFlags = "--with-xkb-config-root=${self.xkb_patched}/share/X11/xkb";
+        });
+
+      };
+
+      ckbcomp = super.ckbcomp.override {
+        xkeyboard_config = self.xkb_patched;
+      };
+
+      xkbvalidate = super.xkbvalidate.override {
+        libxkbcommon = super.libxkbcommon.override {
```

Comment:
I believe this should be `self.libxkbcommon`

Reply:
It doesn't seem to make a difference, though: the build produces the exact same hash.

Reply:
Yeah, but it would make a difference if a different overlay after this one overwrote `libxkbcommon`. With `super` it would not apply, with `self` it would.